corehq/apps/hqwebapp/templatetags/hq_shared_tags.py

```diff
@@ -746,3 +754,14 @@ def request_has_privilege(request, privilege_name):
     from corehq import privileges
     privilege = _get_obj_from_name_or_instance(privileges, privilege_name)
     return domain_has_privilege(request.domain, privilege)
+
+
+@register.filter
+def to_user_time(dt, request):
```

Comment:
nit: definitely the filter is useful. Thinking if this is a good opportunity to name this better?

the function `get_timezone` being called does fallback to domain timezone if user is missing or still use domain's timezone if `override_global_tz` is set to False. So, considering the timezone used could be not that of the user, to_user_time could be misleading.
Could we say `for_user_display` or `to_user_or_domain_timezone`. I do find `get_timezone` could also be named better, like, `get_timezone_for_user_or_domain` or `get_timezone_for_display` but that is out of scope of this PR.

Reply:
I was attempting to mirror the language used in `corehq.util.timezones`, which talks about datetimes as "server time", "user time", or "phone time".  In that context, "server time" is UTC equivalent, but without a timezone attached, and "user time" is a datetime localized to whatever timezone is appropriate to show to the user.

> the function `get_timezone` being called does fallback to domain timezone if user is missing or still use domain's timezone if `override_global_tz` is set to False

I feel like that is an implementation detail detail at a lower abstraction level.  I think `get_timezone` answers the question of "what timezone should we use for this user on this domain"?  That might be a timezone specified on the user, or on the domain, or just UTC, but the answer to that question should be the same in each part of HQ.  So rather than have say the case data page use the user's timezone and the sms logs use the domain timezone, and something else use UTC, they should all use the "user time", as defined by `get_timezone` (and `get_timezone_for_user`, which it calls).

Relatedly, I actually kind of think `get_timezone` should take `request` as its only arg, since you probably want to use the domain from the active request, and it already falls back to UTC if there is no domain.

Reply:
Okay.

I believe I tend to have the naming really explicit and seeing `_user_time` didn't sit well with me because it would at times not be the user time which in my head should be the timezone user had set to and be none or raise error if there was no timezone set and leave for the calling function to figure out what it wants to do, even if it was to fallback to domain timezone.
Though its not always feasible to put everything in the name. So I am not complaining about how its here.

>  I think get_timezone answers the question of "what timezone should we use for this user on this domain"?

Right. I don't prefer to have to look at the function definition to understand what it would be doing or have the name be totally ambiguous so I am forced to look into it. This is in the middle.

> Relatedly, I actually kind of think get_timezone should take request as its only arg, since you probably want to use the domain from the active request, and it already falls back to UTC if there is no domain.

Ya I was unsure in what cases it needed domain to be passed other than the request.

Reply:
> I tend to have the naming really explicit

I've noticed you and I are often on opposite sides of the spectrum here.  I find long names to hurt readability.  I also think of a function's parameter names as part of its public interface, so `get_timezone_from_request(request)` tells me nothing that `get_timezone(request)` doesn't.  I know it's getting it from the request, because I can see it's being passed in.

> it would at times not be the user time

I disagree with this, actually - it's a question of "user" vs `user`.  It will not always be the timezone specified on the `CouchUser`'s `domain_membership`, but it *will* always be timezone that should be shown to the user.  Every time we localize a timestamp to display to the user, we should use the same logic to determine what timezone to use.  There's a quote from _Clean Code_ I'm reminded of:

> Another way to know that a function is doing more than “one thing” is if you can extract another function from it with a name that is not merely a restatement of its implementation

This functionality expresses an abstraction (what timezone should we use), and callers shouldn't need to know about the implementation.  To accurately express the logic happening here it'd have to be something like `get_timezone_from_user_if_global_overridden_else_domain`.  Assuming the name is clear enough to say what the function does, I shouldn't need to think about how it does it when calling it.

Reply:
>  I find long names to hurt readability

same same. If a name needs to be long, the function itself probably can be split.

> I also think of a function's parameter names as part of its public interface

Oh yes, Like i would not mind `get_timezone(user)`, which I would read as `get timezone for user`. So I am aligned with you on this.

> It will not always be the timezone specified on the CouchUser's domain_membership, but it will always be timezone that should be shown to the user

Exactly. And that distinction was what I was getting to with `get_timezone_for_display` to clarify that we are focused on timezone that we should show or display.

> Another way to know that a function is doing more than “one thing” is if you can extract another function from it with a name that is not merely a restatement of its implementation

I like that.

> This functionality expresses an abstraction (what timezone should we use), and callers shouldn't need to know about the implementation

Ya, that makes sense to me, like:

```
def to_user_time(dt, request):
    timezone = _get_timezone_to_use(...)
    return dt.to_timezone(timezone)
```

> To accurately express the logic happening here it'd have to be something like `get_timezone_from_user_if_global_overridden_else_domain`

haha, that would be too much! I might have misspoken about what I like the naming to be.